client/src/instance_spec.rs

```diff
@@ -0,0 +1,287 @@
+    // types to change without causing a breaking change to instance specs.
+    // Consider the following scenario, assuming the VolumeConstructionRequest
+    // struct is used directly:
+    //
+    // - Sled agent v1 starts Propolis v1 using Crucible request v1.
+    // - Sled agent v2 (on some other sled with newer software) starts Propolis
+    //   v2 using Crucible request v2, which has a new field not present in v1.
+    // - Nexus orders a migration from v1 to v2. This requires someone to
+    //   compare the two instances' specs for migratability.
+    //
+    // Migration compatibility is normally checked by the two Propolis servers
+    // involved: one server sends its instance spec to the other, and the
+    // recipient compares the specs to see if they're compatible. In this case,
+    // v2 can't deserialize v1's spec (a field is missing), and v1 can't
+    // deserialize v2's (an extra field is present), so migration will always
+    // fail.
+    //
+    // Storing a serialized request avoids this problem as follows:
+    //
+    // - Sled agent v2 starts Propolis v2 with spec v2. It deserializes the
+    //   request contents in the spec body into a v2 construction request.
+    // - Migration begins. Propolis v2 can now deserialize the v1 instance spec
+    //   and check it for compatibility. It can't deserialize the v1 *request
+    //   contents*, but this can be dealt with separately (e.g. by having the v1
+    //   and v2 Crucible components in the Propolis server negotiate
+    //   compatibility themselves, which is an affordance the migration protocol
+    //   allows).
+    pub json: String,
+}
+
+impl TryFrom<&CrucibleRequestContents> for crucible::VolumeConstructionRequest {
+    type Error = serde_json::Error;
+
+    fn try_from(value: &CrucibleRequestContents) -> Result<Self, Self::Error> {
+        serde_json::from_str(&value.json)
+    }
+}
+
+impl TryFrom<&crucible::VolumeConstructionRequest> for CrucibleRequestContents {
+    type Error = serde_json::Error;
+
+    fn try_from(
+        value: &crucible::VolumeConstructionRequest,
+    ) -> Result<Self, Self::Error> {
+        Ok(Self { json: serde_json::to_string(value)? })
+    }
+}
+
+/// A kind of storage backend: a connection to on-sled resources or other
+/// services that provide the functions storage devices need to implement their
+/// contracts.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub enum StorageBackendKind {
+    /// A Crucible-backed device, containing a generation number and a
+    /// construction request.
+    Crucible { gen: u64, req: CrucibleRequestContents },
+
+    /// A device backed by a file on the host machine. The payload is a path to
+    /// this file.
+    File { path: String },
+
+    /// A device backed by an in-memory buffer in the VMM process. The payload
+    /// is the buffer's contents.
+    InMemory { bytes: Vec<u8> },
```

Comment:
Do we want the actual memory contents of the in-memory disk to be part of its specification here, or simply something that is copied as part of any migration?

Reply:
This is here more for instance init than for migration. In the (admittedly hypothetical) future where `instance_ensure` takes a spec instead of collections of disk, NIC, etc. requests, the spec needs some way to specify what the disk's initial contents should be. (In the current code, doing it this way lets the server initialize this backend only from the spec without having to peek at the disk requests themselves to get this data, which simplifies the code somewhat.)

During state migration, I'd expect the source in-memory backend to save the disk contents for the target to restore (especially if the source disk might've changed). 

In the preamble compatibility check (as I've drafted it), only the backend types matter--the contents are ignored because they can legitimately differ between source and target (e.g. in Crucible generation numbers). That makes it wasteful to serialize and send the disk contents, but I suspect we can deal with that in other ways (e.g. provide a "serialize for compatibility" helper that omits nonessential data).

Reply:
It just seems odd to me to include the contents in the spec.

Reply:
Having examined this more, I think including the contents streamlines the instance initialization code, but the complexity will come back in the form of additional logic to strip them from specs before sending them for migration (which I think I would have to do to look myself in the mirror--I hadn't realized just how much larger the contents are than the rest of the spec). So I've gone ahead and broken this up a bit (and re-tested by creating some local VMs with a cloud-init image).